File: collect_resumes.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

// 浏览器配置 — 使用 Edge
const BROWSER_CHANNEL = 'msedge'; // 可选: 'msedge', 'chrome', 或 null 使用 Chromium

// Cookie数据
const COOKIES = [
    { "domain": ".zhipin.com", "name": "__a", "value": "33417319.1780468874.1780473776.1780542412.24.3.2.2" },
    { "domain": ".zhipin.com", "name": "__c", "value": "1780542412" },
    { "domain": ".zhipin.com", "name": "Hm_lvt_194df3105ad7148dcf2b98a91b5e727a", "value": "1780471580,1780475352,1780542412" },
    { "domain": ".zhipin.com", "name": "wbg", "value": "1", "httpOnly": true },
    { "domain": ".zhipin.com", "name": "__g", "value": "sem_bingpc" },
    { "domain": ".zhipin.com", "name": "zp_at", "value": "SMtdvto0gRqyaujaCoXh6RD7z7Qo-G2yWdqiHta7pzE~", "httpOnly": true },
    { "domain": ".zhipin.com", "name": "lastCity", "value": "101020100" },
    { "domain": ".zhipin.com", "name": "wt2", "value": "Dch7w6ktB3-qwWJQUQVFYF1K5GD3dIzARReaW3pcTGCEQzlvNJlXrWUWpPL3cUJCDYy6Qt64jrmsFjJ08MKz0gw~~", "httpOnly": true },
    { "domain": ".zhipin.com", "name": "__l", "value": "r=https%3A%2F%2Fcn.bing.com%2F&l=%2Fwww.zhipin.com%2Fsem%2F10.html%3F_ts%3D1780542409389%26sid%3Dsem_bingpc%26qudao%3Dbing_pc_H120003UY5%26plan%3DTCPA-%25E5%25BF%2585%25E5%25BA%2594-%25E5%2593%2581%25E7%2589%258C%26unit%3D%25E4%25BD%258E%25E6%2588%2590%25E6%259C%25AC%25E9%25AB%2598%25E6%25B6%2588%25E8%25B4%25B9%25E8%25AF%258D-1215%26keyword%3Dboss%26msclkid%3D4c0374a34ee3133dd0e41e83a5241f8f&s=1&g=%2Fwww.zhipin.com%2Fsem%2F10.html%3F_ts%3D1780542409389%26sid%3Dsem_bingpc%26qudao%3Dbing_pc_H120003UY5%26plan%3DTCPA-%25E5%25BF%2585%25E5%25BA%2594-%25E5%2593%2581%25E7%2589%258C%26unit%3D%25E4%25BD%258E%25E6%2588%2590%25E6%259C%25AC%25E9%25AB%2598%25E6%25B6%2588%25E8%25B4%25B9%25E8%25AF%258D-1215%26keyword%3Dboss%26msclkid%3D4c0374a34ee3133dd0e41e83a5241f8f&s=3&friend_source=0" },
    { "domain": ".zhipin.com", "name": "__zp_seo_uuid__", "value": "8e906530-e1fd-44c0-b9f6-a9811e974db6" },
    { "domain": "www.zhipin.com", "name": "ab_guid", "value": "b7e18443-5aae-4c4e-a420-bed560a8fd06" },
    { "domain": ".zhipin.com", "name": "bst", "value": "V2Sd4nFuT63FpgXdJhzh4fLimw7D3QwQ~~|Sd4nFuT63FpgXdJhzh4fLimw7DnXzQ~~" },
    { "domain": ".zhipin.com", "name": "Hm_lpvt_194df3105ad7148dcf2b98a91b5e727a", "value": "1780542412" },
    { "domain": ".zhipin.com", "name": "HMACCOUNT", "value": "B1187F5BC925506E" }
];

const TARGET_URL = 'https://www.zhipin.com/web/chat/geek/manage_v2';

// 延迟函数
const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

// 随机延迟（模拟人类操作）
const randomDelay = async (min = 800, max = 2000) => {
    const ms = Math.floor(Math.random() * (max - min + 1)) + min;
    await sleep(ms);
};

async function main() {
    console.log('🚀 启动 Boss 直聘收简历机器人（Edge 浏览器）...\n');

    // 检查 Edge 是否可用
    try {
        const { execSync } = require('child_process');
        const edgePath = execSync('where msedge 2>nul || which msedge 2>/dev/null || echo ""').toString().trim();
        if (edgePath) {
            console.log(`✅ Edge 路径: ${edgePath}`);
        } else {
            console.log('⚠️  Edge 未在 PATH 中找到，尝试直接启动...');
        }
    } catch (e) {
        console.log('📌 继续启动 Edge...');
    }

    // 启动 Edge 浏览器（有头模式，便于观察）
    const browser = await chromium.launch({
        headless: false,
        channel: BROWSER_CHANNEL,
        args: ['--disable-blink-features=AutomationControlled']
    });

    const context = await browser.newContext({
        viewport: { width: 1280, height: 800 },
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    });

    // 设置 Cookie
    await context.addCookies(COOKIES);
    const page = await context.newPage();

    let totalRequested = 0;
    let currentPage = 1;
    let hasMorePages = true;

    try {
        console.log('📄 打开页面...');
        await page.goto(TARGET_URL, { waitUntil: 'networkidle', timeout: 30000 });
        await randomDelay(2000, 3000);

        // 循环处理每一页
        while (hasMorePages) {
            console.log(`\n📋 --- 第 ${currentPage} 页 ---`);

            // 等待候选人列表加载
            try {
                await page.waitForSelector('a.btn-startchat, .geek-item, [class*="chat-btn"], button:has-text("沟通")', {
                    timeout: 10000
                });
            } catch (e) {
                console.log('⚠️  未找到候选人列表，可能页面结构已变化');
                await page.screenshot({ path: path.join(__dirname, 'error_screenshot.png') });
                console.log('📸 已保存错误截图: error_screenshot.png');
                break;
            }

            // 收集当前页所有"沟通"按钮
            const chatButtons = await page.$$('a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"]');

            if (chatButtons.length === 0) {
                console.log('⚠️  未找到"沟通"按钮，尝试使用通用选择器...');
                // 尝试更通用的选择器
                const allButtons = await page.$$('button, a');
                const filteredButtons = [];
                for (const btn of allButtons) {
                    const text = await btn.textContent();
                    if (text && text.includes('沟通')) {
                        filteredButtons.push(btn);
                    }
                }
                if (filteredButtons.length === 0) {
                    console.log('⚠️  当前页没有可沟通的候选人');
                    // 检查是否没有更多候选人
                    hasMorePages = await goToNextPage(page, currentPage);
                    currentPage++;
                    continue;
                }
                console.log(`🔍 找到 ${filteredButtons.length} 个"沟通"按钮（通用选择器）`);
                for (let i = 0; i < filteredButtons.length; i++) {
                    await processCandidate(page, i + 1, filteredButtons.length);
                    totalRequested++;
                }
            } else {
                console.log(`🔍 当前页找到 ${chatButtons.length} 个候选人`);

                for (let i = 0; i < chatButtons.length; i++) {
                    await processCandidate(page, i + 1, chatButtons.length);
                    totalRequested++;
                }
            }

            // 切换到下一页
            hasMorePages = await goToNextPage(page, currentPage);
            currentPage++;
            await randomDelay(2000, 3000);
        }

        console.log(`\n✅ 全部完成！共处理 ${totalRequested} 位候选人`);
        console.log('💡 浏览器窗口将保持打开，请手动关闭。');

    } catch (err) {
        console.error('❌ 执行出错：', err.message);
        try {
            await page.screenshot({ path: path.join(__dirname, 'error_screenshot.png') });
            console.log('📸 已保存错误截图: error_screenshot.png');
        } catch (e) { }
    }
}

/**
 * 处理单个候选人：点击沟通 → 求简历 → 确定
 */
async function processCandidate(page, index, total) {
    console.log(`\n👤 处理第 ${index}/${total} 个候选人...`);

    try {
        // 重新获取最新的沟通按钮（避免 stale element）
        await randomDelay(500, 1000);
        const buttons = await page.$$('a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"]');
        if (index - 1 >= buttons.length) {
            console.log('⚠️  按钮已失效，跳过');
            return;
        }

        // 点击"沟通"按钮
        console.log('  点击"沟通"...');
        try {
            await buttons[index - 1].click({ timeout: 5000 });
        } catch (e) {
            // 尝试用 JS 点击
            await page.evaluate((idx) => {
                const btns = document.querySelectorAll('a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"]');
                if (btns[idx]) btns[idx].click();
            }, index - 1);
        }
        await randomDelay(2000, 3000);

        // 等待聊天面板打开，然后点击"求简历"
        console.log('  点击"求简历"...');
        let clickedResume = false;
        try {
            // 尝试多个可能的选择器
            const resumeSelectors = [
                'button:has-text("求简历")',
                'a:has-text("求简历")',
                '[class*="resume"]',
                '[class*="request"]'
            ];
            for (const sel of resumeSelectors) {
                const resumeBtn = await page.$(sel);
                if (resumeBtn) {
                    const text = await resumeBtn.textContent();
                    if (text && text.includes('求简历')) {
                        await resumeBtn.click();
                        clickedResume = true;
                        console.log('  ✅ 已点击"求简历"');
                        break;
                    }
                }
            }
            if (!clickedResume) {
                // 全页面搜索"求简历"
                const allElements = await page.$$('button, a, span, div');
                for (const el of allElements) {
                    try {
                        const text = await el.textContent();
                        if (text && text.trim() === '求简历') {
                            await el.click();
                            clickedResume = true;
                            console.log('  ✅ 已点击"求简历"（通用扫描）');
                            break;
                        }
                    } catch (e) { }
                }
            }
        } catch (e) {
            console.log('  ⚠️  点击"求简历"失败：', e.message);
        }

        if (clickedResume) {
            await randomDelay(1000, 2000);

            // 点击确认对话框的"确定"
            console.log('  确认索取...');
            try {
                const confirmSelectors = [
                    'button:has-text("确定")',
                    'a:has-text("确定")',
                    'button:has-text("确认")',
                    '[class*="confirm"] button:has-text("确定")',
                    '[class*="dialog"] button:has-text("确定")',
                    'button:has-text("好的")'
                ];
                let confirmed = false;
                for (const sel of confirmSelectors) {
                    const confirmBtn = await page.$(sel);
                    if (confirmBtn) {
                        const text = await confirmBtn.textContent();
                        if (text && (text.includes('确定') || text.includes('确认') || text.includes('好的'))) {
                            await confirmBtn.click();
                            confirmed = true;
                            console.log('  ✅ 已点击"确定"');
                            break;
                        }
                    }
                }
                if (!confirmed) {
                    // 在全页面搜索"确定"
                    const allElements = await page.$$('button, a, span, div');
                    for (const el of allElements) {
                        try {
                            const text = await el.textContent();
                            if (text && text.trim() === '确定') {
                                await el.click();
                                confirmed = true;
                                console.log('  ✅ 已点击"确定"（通用扫描）');
                                break;
                            }
                        } catch (e) { }
                    }
                }
                if (!confirmed) {
                    console.log('  ⚠️  未找到"确定"按钮，可能已自动确认');
                }
            } catch (e) {
                console.log('  ⚠️  点击"确定"失败：', e.message);
            }

            await randomDelay(1500, 2500);
        }

        // 关闭聊天面板（点击空白区域或关闭按钮）
        console.log('  关闭聊天面板...');
        try {
            // 尝试点击聊天面板的关闭按钮或背景
            const closeBtn = await page.$('i.icon-close, [class*="close"], button[class*="close"], [class*="dialog"] i');
            if (closeBtn) {
                await closeBtn.click();
            } else {
                // 按 Escape 键关闭
                await page.keyboard.press('Escape');
            }
        } catch (e) {
            await page.keyboard.press('Escape');
        }
        await randomDelay(1000, 1500);

        console.log(`  ✅ 候选人 ${index} 处理完成`);

    } catch (err) {
        console.log(`  ⚠️  处理候选人 ${index} 时出错：`, err.message);
        // 按 Escape 恢复状态
        await page.keyboard.press('Escape').catch(() => { });
        await randomDelay(1000, 2000);
    }
}

/**
 * 切换到下一页
 */
async function goToNextPage(page, currentPageNum) {
    try {
        // 查找分页按钮
        const nextBtn = await page.$('a.next, button.next, [class*="next"], a:has-text("下一页"), button:has-text("下一页")');
        if (!nextBtn) {
            console.log('📄 未找到"下一页"按钮，已在最后一页');
            return false;
        }

        const isDisabled = await nextBtn.evaluate(el => el.classList.contains('disabled') || el.disabled);
        if (isDisabled) {
            console.log('📄 "下一页"按钮已禁用，已到最后一页');
            return false;
        }

        console.log(`📄 切换到第 ${currentPageNum + 1} 页...`);
        await nextBtn.click();
        await randomDelay(2000, 3000);
        await page.waitForLoadState('networkidle');
        return true;

    } catch (e) {
        console.log('📄 切换分页失败：', e.message);
        return false;
    }
}

// 启动主流程
main().catch(console.error);

File: 收简历工具/collect_resumes.py
"""
Boss 直聘批量收简历工具

工作流程：
  1. 脚本自动启动 Edge（调试模式）
  2. 你在打开的 Edge 中手动登录 Boss 直聘，进入「单聊」页面
  3. 脚本自动检测到页面后开始自动化操作
"""

import asyncio
import json
import os
import random
import subprocess
import sys
import time
from datetime import datetime
from pathlib import Path

BASE_DIR = Path(__file__).resolve().parent
TARGET_URL = "https://www.zhipin.com/web/chat/geek/manage_v2"
DEBUG_PORT = 9222
DEBUG_URL = f"http://127.0.0.1:{DEBUG_PORT}"

stats = {"total": 0, "success": 0, "skipped": 0, "errors": 0}


def start_edge_with_debug_port():
    """启动 Edge 调试模式"""
    edge_paths = [
        r"C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe",
        r"C:\Program Files\Microsoft\Edge\Application\msedge.exe",
    ]
    for path in edge_paths:
        if os.path.exists(path):
            print(f"启动 Edge 浏览器（调试端口 {DEBUG_PORT}）...")
            subprocess.Popen(
                [path, f"--remote-debugging-port={DEBUG_PORT}"],
                shell=True,
            )
            return True
    print("未找到 Edge 浏览器")
    return False


async def wait_for_debug_port(timeout=30):
    """等待 Edge 调试端口就绪"""
    import urllib.request
    start = time.time()
    while time.time() - start < timeout:
        try:
            urllib.request.urlopen(DEBUG_URL, timeout=2)
            return True
        except Exception:
            await asyncio.sleep(1)
    return False


async def random_delay(min_ms=800, max_ms=2000):
    ms = random.randint(min_ms, max_ms)
    await asyncio.sleep(ms / 1000)


async def process_candidate(page, index, total_count):
    """处理单个候选人：点击沟通 -> 求简历 -> 确定 -> 关闭"""
    print(f"\n处理第 {index}/{total_count} 个候选人...")
    await random_delay(500, 1000)

    try:
        # --- 点击"沟通" ---
        print("  点击「沟通」...")
        chat_btn = await page.query_selector(
            'a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"]'
        )
        if not chat_btn:
            clicked = await page.evaluate(
                f"""() => {{
                    const btns = document.querySelectorAll('a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"]');
                    if (btns[{index - 1}]) {{ btns[{index - 1}].click(); return true; }}
                    return false;
                }}"""
            )
            if not clicked:
                print("  - 未找到沟通按钮")
                stats["skipped"] += 1
                return
        else:
            await chat_btn.click()
        await random_delay(2000, 3000)

        # --- 点击"求简历" ---
        print("  点击「求简历」...")
        clicked_resume = False
        for selector in [
            'button:has-text("求简历")',
            'a:has-text("求简历")',
            '[class*="resume"]',
            '[class*="request"]',
        ]:
            btn = await page.query_selector(selector)
            if btn:
                text = await btn.inner_text()
                if "求简历" in text:
                    await btn.click()
                    clicked_resume = True
                    print("  - 已点击「求简历」")
                    break

        if not clicked_resume:
            print("  扫描「求简历」...")
            elements = await page.query_selector_all("button, a, span, div")
            for el in elements:
                try:
                    text = await el.inner_text()
                    if text.strip() == "求简历":
                        await el.click()
                        clicked_resume = True
                        print("  - 已点击「求简历」（扫描模式）")
                        break
                except Exception:
                    continue

        if clicked_resume:
            await random_delay(1000, 2000)

            # --- 点击"确定" ---
            print("  确认索取...")
            confirmed = False
            for selector in [
                'button:has-text("确定")',
                'a:has-text("确定")',
                'button:has-text("确认")',
                '[class*="dialog"] button:has-text("确定")',
                'button:has-text("好的")',
            ]:
                btn = await page.query_selector(selector)
                if btn:
                    text = await btn.inner_text()
                    if any(kw in text for kw in ["确定", "确认", "好的"]):
                        await btn.click()
                        confirmed = True
                        print("  - 已点击「确定」")
                        break

            if not confirmed:
                elements = await page.query_selector_all("button, a, span, div")
                for el in elements:
                    try:
                        text = await el.inner_text()
                        if text.strip() == "确定":
                            await el.click()
                            confirmed = True
                            print("  - 已点击「确定」（扫描模式）")
                            break
                    except Exception:
                        continue

            if not confirmed:
                print("  - 未找到「确定」，可能已自动确认")

            await random_delay(1500, 2500)

        # --- 关闭聊天面板 ---
        print("  关闭聊天面板...")
        try:
            close_btn = await page.query_selector(
                'i.icon-close, [class*="close"], button[class*="close"]'
            )
            if close_btn:
                await close_btn.click()
            else:
                await page.keyboard.press("Escape")
        except Exception:
            await page.keyboard.press("Escape")

        await random_delay(1000, 1500)
        print(f"  - 完成")
        stats["success"] += 1

    except Exception as e:
        print(f"  - 出错：{e}")
        stats["errors"] += 1
        try:
            await page.keyboard.press("Escape")
        except Exception:
            pass
        await random_delay(1000, 2000)


async def go_to_next_page(page, current_page_num):
    try:
        next_btn = await page.query_selector(
            'a.next, button.next, [class*="next"], a:has-text("下一页"), button:has-text("下一页")'
        )
        if not next_btn:
            print("未找到「下一页」按钮，已在最后一页")
            return False

        is_disabled = await next_btn.evaluate(
            "el => el.classList.contains('disabled') || el.disabled"
        )
        if is_disabled:
            print("「下一页」按钮已禁用，已到最后一页")
            return False

        print(f"切换到第 {current_page_num + 1} 页...")
        await next_btn.click()
        await random_delay(2000, 3000)
        await page.wait_for_load_state("networkidle")
        return True

    except Exception as e:
        print(f"切换分页失败：{e}")
        return False


async def find_chat_buttons(page):
    buttons = []
    try:
        buttons = await page.query_selector_all(
            'a.btn-startchat, button:has-text("沟通"), [class*="chat-btn"], [class*="startchat"]'
        )
    except Exception:
        pass

    if not buttons:
        try:
            all_el = await page.query_selector_all("button, a, [role='button']")
            for el in all_el:
                try:
                    text = await el.inner_text()
                    if text.strip() in ("沟通", "立即沟通"):
                        buttons.append(el)
                except Exception:
                    continue
        except Exception:
            pass

    if not buttons:
        try:
            data = await page.evaluate("""() => {
                const all = document.querySelectorAll('button, a, li, div[class*="item"], div[class*="card"]');
                const result = [];
                all.forEach((el, i) => {
                    const t = (el.innerText || '').trim();
                    if (t === '沟通' || t === '立即沟通') result.push(i);
                });
                return result;
            }""")
            if data:
                all_el = await page.query_selector_all(
                    "button, a, li, div[class*='item'], div[class*='card']"
                )
                for idx in data:
                    if idx < len(all_el):
                        buttons.append(all_el[idx])
        except Exception:
            pass

    return buttons


async def main():
    print("Boss 直聘批量收简历工具\n")

    # 1. 启动 Edge（调试模式）
    if not start_edge_with_debug_port():
        return

    print("等待浏览器启动...")
    if not await wait_for_debug_port():
        print("Edge 启动超时，请手动以调试模式启动 Edge")
        return

    print("Edge 浏览器已就绪\n")

    from playwright.async_api import async_playwright

    async with async_playwright() as p:
        # 2. 连接 Edge
        print(f"连接 Edge...")
        try:
            browser = await p.chromium.connect_over_cdp(DEBUG_URL)
            print("已连接\n")
        except Exception as e:
            print(f"连接失败：{e}")
            return

        # 3. 获取页面并导航到 Boss 直聘
        page = browser.contexts[0].pages[0] if browser.contexts and browser.contexts[0].pages else await browser.contexts[0].new_page()

        print(f"打开 Boss 直聘（{TARGET_URL}）...")
        try:
            await page.goto(TARGET_URL, wait_until="domcontentloaded", timeout=30000)
        except Exception:
            pass

        # 检查是否跳到登录页
        await asyncio.sleep(2)
        current_url = page.url
        found = False

        if "login" in current_url or "passport" in current_url:
            print("=" * 60)
            print("请在打开的 Edge 浏览器中登录 Boss 直聘")
            print("登录后将自动进入单聊页面，脚本会自动检测")
            print("=" * 60)
        elif "chat" in current_url or "geek" in current_url:
            # 已登录且已进入目标页面
            body = await page.evaluate("() => document.body.innerText.substring(0, 2000)")
            if "沟通" in body:
                print("已进入单聊页面，开始处理...\n")
                found = True

        # 4. 如果还没就绪，开始等待
        if not found:
            print("脚本每 2 秒自动检测，检测到单聊页面后立即开始...\n")
            for wait_round in range(150):  # 5 分钟
                await asyncio.sleep(2)
                try:
                    url = page.url
                    body = await page.evaluate("() => document.body.innerText.substring(0, 5000)")
                    url_ok = any(kw in url for kw in ["web/chat", "geek", "manage"])
                    text_ok = "沟通" in body
                    if url_ok and text_ok:
                        found = True
                        print("已检测到单聊页面，开始处理...\n")
                        break
                    if wait_round % 15 == 0:
                        short_url = url[:60] if len(url) > 60 else url
                        print(f"等待中（{wait_round * 2}s）... {short_url}")
                except Exception:
                    if wait_round % 15 == 0:
                        print("页面加载中...")

        if not found:
            print("等待超时，请重启脚本")
            return

        # 5. 主循环
        current_page_num = 1
        has_more = True

        while has_more:
            print(f"\n{'='*50}")
            print(f"第 {current_page_num} 页")
            print(f"{'='*50}")

            await asyncio.sleep(2)

            chat_buttons = await find_chat_buttons(page)

            if not chat_buttons:
                print("当前页没有「沟通」按钮")
                try:
                    await page.screenshot(path=str(BASE_DIR / f"page_{current_page_num}.png"))
                except Exception:
                    pass
                has_more = await go_to_next_page(page, current_page_num)
                current_page_num += 1
                continue

            total_candidates = len(chat_buttons)
            stats["total"] += total_candidates
            print(f"找到 {total_candidates} 个候选人")

            for i in range(total_candidates):
                await process_candidate(page, i + 1, total_candidates)

            has_more = await go_to_next_page(page, current_page_num)
            current_page_num += 1

        print(f"\n{'='*50}")
        print("全部完成！")
        print(f"{'='*50}")
        print(f"共处理：{stats['total']}")
        print(f"成功：{stats['success']}")
        print(f"跳过：{stats['skipped']}")
        print(f"出错：{stats['errors']}")


if __name__ == "__main__":
    asyncio.run(main())